build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -2000,6 +2188,42 @@ message Compressor {
   }
 }
 
+// Content-defined chunking algorithms used for splitting blobs into chunks.
+message ChunkingAlgorithm {
+  enum Value {
+    // No special constraints about the used chunking algorithm. If a client
+    // selects this value this means he does not care about which exact chunking
+    // algorithm is used. Thus, the server is free to use any chunking algorithm
+    // at its disposal. A server does not need to advertise this value.
+    DEFAULT = 0;
+
+    // Content-defined chunking using Rabin fingerprints. An implementation of
+    // this scheme in presented in this paper
+    // https://link.springer.com/chapter/10.1007/978-1-4613-9323-8_11. The final
+    // implementation of this algorithm should be configured to have the
+    // following properties on resulting chunk sizes.
+    //  - Minimum chunk size: 128 KB
+    //  - Average chunk size: 512 KB (0x000000000007FFFF bit mask)
+    //  - Maximum chunk size: 2048 KB
+    // The irreducible polynomial to be used for the modulo divisions is the
+    // following 64-bit polynomial of degree 53: 0x003DA3358B4DC173. The window
+    // size to be used is 64 bits.
+    RABINCDC = 1;
+
+    // Content-defined chunking using the FastCDC algorithm. The algorithm is
+    // described in this paper https://ieeexplore.ieee.org/document/9055082
+    // (Algorithm 2, FastCDC8KB). The algorithm should be configured to have the
+    // following properties on resulting chunk sizes.
+    //  - Minimum chunk size: 128 KB
+    //  - Average chunk size: 512 KB
+    //  - Maximum chunk size: 2048 KB
```

Comment:
Looking at algorithm 2 in that paper, I see that I can indeed plug in these values to MinSize, MaxSize, and NormalSize. But what values should I use for MaskS and MaskL now? (There's also MaskA, but that seems unused by the algorithm.)

Reply:
Good question, as far as I have understood the algorithm, it is the normalized chunking technique that they use, which allows to keep these mask values as they are, but change the min/max/average chunk sizes according to your needs.

Reply:
Looking at algorithm 2 in that paper that attempts to compute 8 KB chunks, I think the probability of a random chunk having size `x` should be as follows ([link](https://www.desmos.com/calculator/2d2gqajtqb)):

![desmos-graph](https://github.com/bazelbuild/remote-apis/assets/736085/65d26cf6-d52a-47e0-bb7e-93fbd246db31)

Now if I change that graph to use the minimum/average/maximum chunk sizes that you propose while leaving the bitmasks unaltered, I see this ([link](https://www.desmos.com/calculator/h92klnucmo)):

![desmos-graph-2](https://github.com/bazelbuild/remote-apis/assets/736085/a05425f2-dde3-4ab0-9c86-3972474064fb)

If I change it so that `MaskS` has 21 bits and `MaskL` has 17, then the graph starts to resemble the original one ([link](https://www.desmos.com/calculator/zpuf6rhubw)):

![desmos-graph-3](https://github.com/bazelbuild/remote-apis/assets/736085/40c55d29-86a8-4ae9-8208-5fb162837146)

So I do think the masks need to be adjusted as well.

Reply:
Hmm... Making the graphs above was meaningful. I think it gives good insight in why FastCDC was designed the way it is. They are essentially trying to mimic a normal distribution ([link](https://www.desmos.com/calculator/v91cuknzrs)):

![desmos-graph-4](https://github.com/bazelbuild/remote-apis/assets/736085/8b6f9fdc-6917-4ef9-a6bf-c4cd808a7abc)

Using a literal normal distribution would not be desirable, because it means that the probability that a chunk is created at a given point not only depends on the data within the window, but also the size of the current chunk. And this is exactly what CDC tries to prevent. So that's why they emulate it using three partial functions.

Yeah, we should make sure to set `MaskS` and `MaskL` accordingly.

Reply:
Wow, thanks @EdSchouten for this great analysis. I have to admit, I did not look into these mask values in that detail, because it also was not really explained in detail in the paper, but your concerns are absolutely right and we have to set the mask values accordingly, when the min/max/average chunk sizes are changed. I am just asking myself, since the paper authors mentioned they derived the mask values empirically, how should we adapt them?

@luluz66 , @sluongng since you mentioned you were working with the FastCDC algorithm with chunk sizes of 500 KB, I am wondering how you handled the adaption of the mask values or whether you did it at all. Can you share your experience here? Thank you so much.

Reply:
We have our fork of CDC implementation here https://github.com/buildbuddy-io/fastcdc-go/blob/47805a2ecd550cb875f1b797a47a1a648a1feed1/fastcdc.go#L162-L172

This work is very much in progress and not final. We hope that the current API will come with configurable knobs so that downstream implementation could choose what's best for their use cases and data.

Reply:
> I am just asking myself, since the paper authors mentioned they derived the mask values empirically, how should we adapt them?

I think that the authors of the paper just used the following heuristics:

1. Don't use the bottom 16 bits, because those likely contain low quality data.
2. Don't use the top 16 bits to get an equal comparison against many common Rabin fingerprinting implementations that use a 48 byte window.

(1) sounds like a good idea, but (2) can likely be ignored. So just pick 21 out of the top 48 bits, and then another 17 bits that area subset of the former.